lib/index.js

```
@@ -24,287 +25,227 @@ const internals = {
 };
 
 
-exports.parse = function (req, tap, options, next) {
+exports.parse = async function (req, tap, options) {
 
     Hoek.assert(options, 'Missing options');
     Hoek.assert(options.parse !== undefined, 'Missing parse option setting');
     Hoek.assert(options.output !== undefined, 'Missing output option setting');
 
-    const parser = new internals.Parser(req, tap, options, next);
-    return parser.read();
-};
-
-
-internals.Parser = function (req, tap, options, next) {
-
-    this.req = req;
-    this.settings = options;
-    this.tap = tap;
-
-    this.result = {};
-    this.next = (err) => next(err, this.result);
-};
-
-
-internals.Parser.prototype.read = function () {
-
-    const next = this.next;
-
     // Content size
 
-    const req = this.req;
     const contentLength = req.headers['content-length'];
-    if (this.settings.maxBytes !== undefined &&
+    if (options.maxBytes !== undefined &&
         contentLength &&
-        parseInt(contentLength, 10) > this.settings.maxBytes) {
+        parseInt(contentLength, 10) > options.maxBytes) {
 
-        return next(Boom.entityTooLarge('Payload content length greater than maximum allowed: ' + this.settings.maxBytes));
+        throw Boom.entityTooLarge('Payload content length greater than maximum allowed: ' + options.maxBytes);
     }
 
     // Content type
 
-    const contentType = Content.type(this.settings.override || req.headers['content-type'] || this.settings.defaultContentType || 'application/octet-stream');
+    const contentType = Content.type(options.override || req.headers['content-type'] || options.defaultContentType || 'application/octet-stream');
     if (contentType.isBoom) {
-        return next(contentType);
+        throw contentType;
```

If `content` is updated to throw errors instead of returning them, this if case can go away.

---

Yep. Part of the final wave of refactoring. For now I am focused on hapi core and only moving modules I get blocked on...